### PR TITLE
adds certificate pinning in the certificate validation process

### LINKED ISSUE #3556: Can't send email using Office365 account
SMTP client fails due to missing automatic migration

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
adds a CertificateThumbnail setting. When used, it will check the server's certificate with this value to decide if the certificate is valid or not.

fixes #3556

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Email/Services/SmtpService.cs [hunk 1]
```diff
@@ -125,7 +122,33 @@ private MimeMessage FromMailMessage(MailMessage message)
...
+        {
+            bool valid = sslPolicyErrors == SslPolicyErrors.None;
+            if (!string.IsNullOrWhiteSpace(_options.CertificateThumbprint))
+            {
+                valid = certificate.GetCertHashString().Equals(_options.CertificateThumbprint);
+            }
+
+            if (!valid)
+            {
+                _logger.LogError($"SMTP Server's certificate {certificate.Subject} issued by {certificate.Issuer} " +
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Email/Services/SmtpService.cs]: Logging shouldn't be used with interpolated strings. It handles `{argument}` by itself. Please take a look at other usages in OC

File: src/OrchardCore.Modules/OrchardCore.Email/Views/SmtpSettings.Edit.cshtml [hunk 2]
```diff
@@ -79,6 +79,13 @@
                     </div>
                 </div>
 
+                <div class="form-group" asp-validation-class-for="CertificateThumbprint">
+                    <label asp-for="CertificateThumbprint">@T["SMTP Server's certificate Thumbprint"]</label>
+                    <input asp-for="CertificateThumbprint" class="form-control" type="text" />
+                    <span asp-validation-for="CertificateThumbprint"></span>
+                    <span class="hint">@T["Pin the SMTP server's certificate. If used and the certificate has different Thumbprint, the certificate will be considered invalid. Check log file for entries regarding the certificate validation process."]</span>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Email/Views/SmtpSettings.Edit.cshtml]: Shouldn't it be written the other way: "When used, the certificate will be considered valid if the thumbprint matches this value"

File: src/OrchardCore.Modules/OrchardCore.Email/Services/SmtpService.cs [hunk 3]
```diff
@@ -132,16 +132,19 @@ private bool CertificateValidationCallback(object sender, X509Certificate certif
...
+                    certificate.Subject, certificate.Issuer, certificate.GetCertHashString(),
+                    certificate.GetExpirationDateString(), sslPolicyErrors);
             }
 
             if (sslPolicyErrors.HasFlag(SslPolicyErrors.RemoteCertificateChainErrors))
             {
                 for (int i = 0; i < chain.ChainStatus.Length - 1; i++)
                 {
-                    _logger.LogError($"{i} = {chain.ChainStatus[i].StatusInformation} : {chain.ChainStatus[i].Status}");
+                    _logger.LogError("chainStatus[{ChainStatusIndexi}] = {Status}:{StatusInformation} ",
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Email/Services/SmtpService.cs]: I don't see the value of including the index. Just having them all outputed should be fine. Using a custom column for the index feels dirty to me.

**Comment by MichaelPetrinolis** [comment]
i do not check chain and ChainStatus for null, i believe that if SslPolicyErrors.RemoteCertificateChainErrors flag is set , it should also set the chain status.

**Comment by sebastienros** [comment]
Not sure we need this now that the issue was found. What do you think?

**Comment by MichaelPetrinolis** [comment]
I agree, smtp server should have a valid certificate. I will create a new PR to add just the logging for certificate validation errors.

**Pull request closed without merging**